nixosTests.networking: refactor and add NetworkManager support

Janik-Haag opened the pull request:
## Description of changes

This adds nixos vm tests for NetworkManager and simplifies the existing  networking tests for systemd-networkd and scripted networking. There is still a lot of todo's and this is a early-ish draft for now.

Fixes #71151

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Janik-Haag committed:
nixosTests.networking: refactor and add NetworkManager support

Janik-Haag committed:
nixos/networkmanager: change config generation to use the ini generator

Janik-Haag committed:
networkmanager: add passthru.tests

samueldr commented on pkgs/tools/networking/networkmanager/default.nix:
```diff
@@ -203,6 +204,9 @@ stdenv.mkDerivation rec {
       attrPath = "networkmanager";
       versionPolicy = "odd-unstable";
     };
+    tests ={
```
```suggestion
    tests = {
```

samueldr commented on nixos/tests/networking/networkmanager.nix:
```diff
@@ -0,0 +1,153 @@
+        static.wait_until_fails("cat /etc/resolv.conf | grep -q '192.168.1.1'")
+      '';
+    };
+    # wireguard = {
+    # };
+    # dispatcherScripts = {
+    # };
```
Should this be removed? Implemented?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/45

bjornfor commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -361,6 +361,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
 - The `hardware.pulseaudio` module now sets permission of pulse user home directory to 755 when running in "systemWide" mode. It fixes [issue 114399](https://github.com/NixOS/nixpkgs/issues/114399).
 
+- The `services.networkmanager.extraConfig` option was changed to use the ini type instead of using a multiline string, this is a breaking change.
```
Isn't this the type of config we call "settings" now? Any reason we're not using that name here?

Janik-Haag replied:
not really, would probably be correcter thing now since settings is the name we give rfc42 compliant options. I can add a `lib.mkRemovedOptionModule` error for extraConfig and rename it to settings.

Janik-Haag replied:
fixed

Janik-Haag committed:
codeowners: add Janik to NetworkManager module and tests

Janik-Haag commented:
I think this is now mergable. 
(I'll squash all the fixup commits once this was approved because if I do it before hand and force push reviewing gets annoying because github resets the interface on force push's)

jtojnar commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -361,6 +361,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
 - The `hardware.pulseaudio` module now sets permission of pulse user home directory to 755 when running in "systemWide" mode. It fixes [issue 114399](https://github.com/NixOS/nixpkgs/issues/114399).
 
+- The `services.networkmanager.extraConfig` was renamed to `services.networkmanager.extraConfig` and was changed to use the ini type instead of using a multiline string.
```
Both names the same.

Janik-Haag replied:
whoops, fixed the copy past error.

lf- commented on nixos/modules/services/networking/networkmanager.nix:
```diff
@@ -454,8 +436,16 @@ in
+      [ "networking" "networkmanager" "useDnsmasq" ]
+      [ "networking" "networkmanager" "dns" ]
+    )
+    (mkRenamedOptionModule
```
hmmm i don't think it's quite a rename, this is a semantic change to be an attr set that should have a nicer message and probably stop evaluation saying how to fix it.

Janik-Haag replied:
Can you share a example of how to do that or point me to the correct function?

Ma27 replied:
`mkRemovedOptionModule` with a good description?
My expectation would be otherwise that I can rename the option in my config and call it a day.

That said, perhaps it's good to keep `extraConfig` for one more release?

Janik-Haag replied:
> That said, perhaps it's good to keep extraConfig for one more release?

I can almost certainly guarantee you that everyone will just forgot about it until then, so I would either do it now or leave it as is, but I don't have hard opinions on this.

> mkRemovedOptionModule with a good description?

sounds good to me.

Ma27 replied:
> I can almost certainly guarantee you that everyone will just forgot about it until then

I mean, it should give an annoying warning to make sure that this won't happen ;)

Janik-Haag replied:
I changed it to `mkRemovedOptionModule` (not sure if the markdown works there).
> > I can almost certainly guarantee you that everyone will just forgot about it until then

> I mean, it should give an annoying warning to make sure that this won't happen ;)

I meant if we do not rename extraConfig to settings that no body will remember to rename it later.
But I probably just misunderstood: "That said, perhaps it's good to keep extraConfig for one more release?"

lf- commented on nixos/tests/networking/networkd-and-scripted.nix:
```diff
@@ -139,35 +57,23 @@ let
           with subtest("Test default gateway"):
-              router.wait_until_succeeds("ping -c 1 192.168.3.1")
               client.wait_until_succeeds("ping -c 1 192.168.3.1")
-              router.wait_until_succeeds("ping -c 1 fd00:1234:5678:3::1")
```
not fully sure where this moved, so unsure what changed here.

Janik-Haag replied:
In my opinion it doesn't make sense for a machine to ping it self if the other thest vm is able to communicate with the corresponding ip so I removed the lines where that was done.

lf- commented on nixos/tests/networking/networkd-and-scripted.nix:
```diff
@@ -300,26 +184,23 @@ let
             driverOptions.mode = "802.3ad";
           };
-          interfaces.bond0.ipv4.addresses = mkOverride 0
+          interfaces.bond0.ipv4.addresses = lib.mkOverride 0
```
would it be possible to move these `with lib` changes into a separate pr that can be insta merged? or at least a separate commit? (can use git-revise, jujutsu, or similar to do that semi easily)

Janik-Haag replied:
see #296500

lf- reviewed (commented):
some review done, thanks for writing these!

there's some of the refactoring stuff that could be split into "move stuff" and "change stuff" to make review a little bit easier.

lf- reviewed (approved):
thanks for writing this~

Aleksanaa commented on nixos/modules/services/networking/networkmanager.nix:
```diff
@@ -169,11 +151,11 @@ in
 
-      extraConfig = mkOption {
-        type = types.lines;
-        default = "";
```
I don't know if this is a de facto standard, but even if many modules are moved to `settings`, often they also retain `extraConfig`. Some new modules are still adding `extraConfig`.

The benefit of this is obvious: we cannot predict that we will always generate configs for users that meet their exact needs. Some exceptions may occur inadvertently, so a guaranteed option is always good.

Janik-Haag replied:
Changing to settings is the preferred key if every value can be defined using nix data structures and it's the "RFC42 style". Changing the name also doesn't really matter here since the option would break either way because the datatype was changed.

lf- replied:
it would be possible to say, have extraConfig that gets smacked on the very end of the file, which is what `nix.settings` did, in spite of having structured config as the primary method, to allow for migration. Personally I am fine either way, the only real argument to not breaking it immediately and hard is to give time to migrate (especially for any downstream modules that need to be cross compatible across NixOS versions), but I don't think this is terrifically critical.

Janik-Haag replied:
Agreed + parsing a ini and converting it to nix is fairly trivial for any usecase where the config is more then like 5 lines, I basically automated that in https://github.com/Janik-Haag/nm2nix/pull/6

lf- commented:
fixes: https://github.com/NixOS/nixpkgs/issues/71151

Janik-Haag commented on nixos/tests/networking/networkmanager.nix:
I don't think so since we I didn't write a testcase where a specific service tries to reach something on the remote. But it sounds reasonable to add something like this.

Janik-Haag commented on nixos/tests/networking/networkmanager.nix:
```diff
@@ -0,0 +1,169 @@
+        dynamic.wait_for_unit("NetworkManager.service")
+        static.wait_for_unit("NetworkManager.service")
+
+        dynamic.wait_until_succeeds("cat /etc/resolv.conf | grep -q '192.168.1.1'")
+        static.wait_until_succeeds("cat /etc/resolv.conf | grep -q '10.10.10.10'")
+        static.wait_until_fails("cat /etc/resolv.conf | grep -q '192.168.1.1'")
```
Uhm sure we can also add a test case for checking if dns actually works. This test case was born out of #291868 which motivated me to write NM tests because I really did not want to manually test stuff every time someone opens a issue related to the NM module.

Janik-Haag commented on nixos/tests/networking/networkd-and-scripted.nix:
yeah the changes there should be identical and disappear once #296500 is merged and I can rebase this pr.

Janik-Haag replied:
I rebased now that the other pr is merged and the changes are gone so I'll mark this as resolved.

Janik-Haag commented on nixos/modules/services/networking/networkmanager.nix:
```diff
@@ -454,8 +436,28 @@ in
+      -      [main]
+      -      no-autodefault=*
+      -    '''
+      +    extraConfig.main.no-auto-default = "*";
```
fixed

lilyinstarlight reviewed (approved):
Diff looks good to me! I see there is an open review comment about end-to-end DNS checks, but that can probably be done later in a follow-up so this PR can go ahead and get merged

I'll merge it when ofborg is green

Janik-Haag commented:
> Diff looks good to me! I see there is an open review comment about end-to-end DNS checks, but that can probably be done later in a follow-up so this PR can go ahead and get merged

yes, there is two open comments about more tests:
- https://github.com/NixOS/nixpkgs/pull/292472#discussion_r1543904457
- https://github.com/NixOS/nixpkgs/pull/292472#discussion_r1543899085

but I want to add them in another pr because I want to get this in before 24.05 branch-of (because of the NM module changes)

> I'll merge it when ofborg is green

thank you <3
you can also build `nix-build -A nixosTests.networking.*` if you want to be 100% sure (afaik ofborg only builds vm tests if a package with passthru.tests get's modified)

lf- commented on nixos/tests/networking/networkmanager.nix:
```diff
@@ -0,0 +1,169 @@
+      };
+      testScript = ''
+        start_all()
+        router.wait_for_unit("network-online.target")
```
oh gods. i am sorry i didn't notice this earlier. this is broken when `network-online.target` is not depended on by multi-user.target.
```suggestion
        router.systemctl("start network-online.target")
        router.wait_for_unit("network-online.target")
```

Janik-Haag replied:
I think it's fixed now. Thank's for the headsup!

lf- commented on nixos/tests/networking/networkmanager.nix:
```diff
@@ -0,0 +1,169 @@
+      };
+      testScript = ''
+        start_all()
+        router.wait_for_unit("network-online.target")
```
```suggestion
        router.systemctl("start network-online.target")
        router.wait_for_unit("network-online.target")
```

lf- commented on nixos/tests/networking/networkmanager.nix:
```diff
@@ -0,0 +1,169 @@
+      };
+      testScript = ''
+        start_all()
+        router.wait_for_unit("network-online.target")
```
```suggestion
        router.systemctl("start network-online.target")
        router.wait_for_unit("network-online.target")
```

Janik-Haag committed:
nixosTests.networking: start network-online.target manually

this is done to avoid future problems for when `network-online.target` doesn't depend on `multi-user.target`

lf- commented:
Builds on my machine, and the changes are trivial enough to trust the last ofborg run.  Merging.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/is-gnome-supposed-to-detect-captive-portals/44417/7
